### DISCUSSION TITLE
why use a 555 at all when ESP can generate required squarewave?

### DISCUSSION THREAD

**Discussion by ll3141** [PricelessToolkit/SOILSENS-V5W#2] [General]
why do you use the 555 at all, when you have an ESP on the board?
The ESP can easyly generate the squarewave required for the sensor.

**Comment by PricelessToolkit** [comment]
There are a couple of reasons!

- I originally started with the 555 timer, and by the time everything was nearly set up, I only then considered using PWM.
- I was concerned it might not work stably with WiFi running at the same time.
- I’m not exactly a pro programmer, so with my janky code, I didn’t want to risk messing up the PWM timings and "moisture values".
- The 555 timer seemed more stable in this setup.

Hope that explains it well! 😊

**Reply by ll3141** [reply]
I'm not complaining 8-)
just curiously asking because I've yet to see a capacitive moisture sensor with integrated ESP without the 555 and am asking myself why that is.

The ESP32 has hardware PWM units (LEDC & MCPWM) so frequency/duty cycle stability should be no problem

Using the ESP-PWM to generate the squarewave could be beneficial for low moisture value by reducing the frequency or shifting the duty cycle to increase integration time.
